sys/cheri/revoke.h

```diff
@@ -0,0 +1,476 @@
+
+#include <cheri/cherireg.h> // CHERI_OTYPE_BITS
+
+typedef uint64_t cheri_revoke_epoch;
+#define CHERI_REVOKE_ST_EPOCH_WIDTH	61
+
+#ifdef _KERNEL
+/*
+ * The outermost capability revocation state machine.
+ *
+ *   We begin with no revocation in progress (CAPREVST_NONE).  Userland can
+ *   request that we steal the calling thread to carry out an initial pass,
+ *   which will begin the next epoch and visit all cap-dirty pages in the
+ *   address space, moving us to CAPREVST_INIT_PASS for the duration and to
+ *   CAPREVST_INIT_DONE upon completion, before returning to userland.
+ *
+ *   We are now in the store-side steady state for this epoch: we need only
+ *   look at *recently* capdirty pages, kernel hoarders, and register files
+ *   to find should-be-revoked pages.  Userland can request additional
+ *   passes, which will be *incremental* passes, visiting only capdirty
+ *   pages (and keeping us in CAPREVST_STORE_DONE).
+ *
+ *   Alternatively, userland can request that we finish this epoch,
+ *   transitioning us into CAPREVST_LAST_PASS for the duration.  This pass
+ *   visits all recently capdirty pages, kernel hoarders, and thread
+ *   register files.  It implies at least some brief window of
+ *   thread_single'd life.  When this pass finishes, it again increments
+ *   the epoch counter, signaling the end of this epoch.
+ *
+ *     XXX The thread_single window can be reduced by using the "load-side"
+ *     strategy, in which we temporarily remove read access to all
+ *     recently-capdirty pages and then un-thread_single.  We would continue
+ *     to use the calling thread to sweep pages "in the background", but would
+ *     also clean pages when taking vm faults.  The epoch must not end
+ *     before all pages have been cleaned!
+ *
+ *   Userland might cause us to jump directly from CAPREVST_NONE to
+ *   CAPREVST_LAST_PASS.  This case is *almost* like the above, except that
+ *   we have to consider *all* capdirty pages, not just recently-capdirty.
+ *
+ * The state and epoch counter are stored in the same per-process word,
+ * vm_caprev_st.  There is, at present, at most one thread actively engaged
+ * in revocation per process.
+ */
+
+enum cheri_revoke_state {
+	CHERI_REVOKE_ST_NONE       = 0, /* No revocation is in progress */
+	CHERI_REVOKE_ST_SS_INITING = 1, /* "store-side" opening now */
+	CHERI_REVOKE_ST_SS_INITED  = 2, /* "store-side" open (> 0 opens done) */
+	CHERI_REVOKE_ST_SS_LAST    = 3, /* "store-side" closing */
+	CHERI_REVOKE_ST_LS_INITING = 4, /* "load-side" opening now */
+	CHERI_REVOKE_ST_LS_INITED  = 5, /* "load-side" open (= 1 opens done) */
+	CHERI_REVOKE_ST_LS_CLOSING = 6, /* "load-side" background working */
+};
+
+#define CHERI_REVOKE_ST_ST_MASK	0x7
+#define CHERI_REVOKE_ST_EPOCH_SHIFT	3
+
+static inline enum cheri_revoke_state
+cheri_revoke_st_state(uint64_t st) {
+	return (st & CHERI_REVOKE_ST_ST_MASK);
+}
+
+static inline uint64_t
+cheri_revoke_st_epoch(uint64_t st) {
```

Comment:
I've renamed these functions and moved these functions out to a different header so that they are not part of the initial merge.  They're not meaningful to userspace anyway.